lib/absinthe/type/union.ex

```diff
@@ -75,16 +75,11 @@ defmodule Absinthe.Type.Union do
   def resolve_type(type, object, env, opts \\ [lookup: true])
   def resolve_type(%{resolve_type: nil, types: types}, obj, %{schema: schema}, opts) do
     type_name = Enum.find(types, fn
-      %{is_type_of: nil} ->
-        false
+      %{is_type_of: nil} -> false
       type ->
         case Schema.lookup_type(schema, type) do
-          nil ->
-            false
-          %{is_type_of: nil} ->
```

Comment:
`is_type_of` will never be nil?

Reply:
Yep, Dialyzer issued a large error for it.

Reply:
```erlang
lib/absinthe/type/union.ex:84: The pattern #{'is_type_of':='nil'} can never match the type #{'__struct__':='Elixir.Absinthe.Type.Enum' | 'Elixir.Absinthe.Type.Field' | 'Elixir.Absinthe.Type.InputObject' | 'Elixir.Absinthe.Type.Interface' | 'Elixir.Absinthe.Type.List' | 'Elixir.Absinthe.Type.NonNull' | 'Elixir.Absinthe.Type.Object' | 'Elixir.Absinthe.Type.Scalar' | 'Elixir.Absinthe.Type.Union', '__private__'=>[{_,_}], '__reference__'=>#{'__struct__':='Elixir.Absinthe.Type.Reference', 'identifier':=atom(), 'module':=atom(), 'name':=binary()}, 'args'=>'nil' | #{atom() | binary()=>map()}, 'complexity'=>'nil' | fun((...) -> non_neg_integer()) | non_neg_integer() | {atom(),atom()}, 'config'=>_, 'default_value'=>_, 'deprecation'=>'nil' | #{'reason':=binary()}, 'description'=>'nil' | binary(), 'field_imports'=>_, 'fields'=>fun(() -> map()) | map(), 'identifier'=>atom(), 'interfaces'=>[map()], 'is_type_of'=>fun((_) -> boolean()), 'middleware'=>[], 'name'=>binary(), 'of_type'=>_, 'parse'=>fun((_) -> 'error' | {_,_}), 'resolve_type'=>fun((_,map()) -> atom()), 'serialize'=>fun((_) -> any()), 'triggers'=>_, 'type'=>atom(), 'types'=>[atom()], 'values'=>#{binary()=>map()}, 'values_by_internal_value'=>_, 'values_by_name'=>_}
```

Reply:
Given this schema:
```
defmodule Foo do
  use Absinthe.Schema

  query do

  end

  union :foo do
    types [:user, :dog]
  end

  object :user do
  end

  object :dog do
  end

end
```

```
iex(2)> Absinthe.Schema.lookup_type(Foo, :foo)
%Absinthe.Type.Union{__private__: [],
 __reference__: %{identifier: :foo, location: %{file: "iex", line: 8},
   module: Foo}, description: nil, name: "Foo", resolve_type: nil,
 types: [:user, :dog]}
iex(3)> Absinthe.Schema.lookup_type(Foo, :user)
%Absinthe.Type.Object{__private__: [],
 __reference__: %{identifier: :user, location: %{file: "iex", line: 12},
   module: Foo}, description: nil, field_imports: [], fields: %{},
 identifier: :user, interfaces: [], is_type_of: nil, name: "User"}
```

resolve_type is nil on the union, and is_type_of is nil on the object type.

Reply:
Sounds like we need to update a spec then....

Reply:
I've removed this change entirely since it sounds more like it's the wrong solution to the problem.

Reply:
yeah it looks like the real issue is that the output of `lookup_types` should really be one of the type structs, not `Type.t` which merely refers (at the moment at least) to one of the type modules, not to their structs.

Reply:
actually no that isn't true.

Reply:
ah, here may be the issue: https://github.com/absinthe-graphql/absinthe/blob/v1.3.2/lib/absinthe/type/object.ex#L93 should be `| nil`

Reply:
Yeah, that could very well be the issue.